operator/pkg/clustermesh/endpointslice_controller.go

```diff
@@ -0,0 +1,607 @@
+	podLister corelisters.PodLister
+	// podsSynced returns true if the pod shared informer has been synced at least once.
+	// Added as a member to the struct to allow injection for testing.
+	podsSynced cache.InformerSynced
+
+	// endpointSliceLister is able to list/get endpoint slices and is populated by the
+	// shared informer passed to NewController
+	endpointSliceLister discoverylisters.EndpointSliceLister
+	// endpointSlicesSynced returns true if the endpoint slice shared informer has been synced at least once.
+	// Added as a member to the struct to allow injection for testing.
+	endpointSlicesSynced cache.InformerSynced
+	// endpointSliceTracker tracks the list of EndpointSlices and associated
+	// resource versions expected for each Service. It can help determine if a
+	// cached EndpointSlice is out of date.
+	endpointSliceTracker *endpointsliceutil.EndpointSliceTracker
+
+	// nodeLister is able to list/get nodes and is populated by the
+	// shared informer passed to NewController
+	nodeLister corelisters.NodeLister
+	// nodesSynced returns true if the node shared informer has been synced at least once.
+	// Added as a member to the struct to allow injection for testing.
+	nodesSynced cache.InformerSynced
+
+	// reconciler is an util used to reconcile EndpointSlice changes.
+	reconciler *endpointslicerec.Reconciler
+
+	// triggerTimeTracker is an util used to compute and export the
+	// EndpointsLastChangeTriggerTime annotation.
+	triggerTimeTracker *endpointsliceutil.TriggerTimeTracker
+
+	// Services that need to be updated. A channel is inappropriate here,
+	// because it allows services with lots of pods to be serviced much
+	// more often than services with few pods; it also would cause a
+	// service that's inserted multiple times to be processed more than
+	// necessary.
+	queue workqueue.RateLimitingInterface
+
+	// maxEndpointsPerSlice references the maximum number of endpoints that
+	// should be added to an EndpointSlice
+	maxEndpointsPerSlice int32
+
+	// workerLoopPeriod is the time between worker runs. The workers
+	// process the queue of service and pod changes
+	workerLoopPeriod time.Duration
+
+	// endpointUpdatesBatchPeriod is an artificial delay added to all service syncs triggered by pod changes.
+	// This can be used to reduce overall number of all endpoint slice updates.
+	endpointUpdatesBatchPeriod time.Duration
+
+	// topologyCache tracks the distribution of Nodes and endpoints across zones
+	// to enable TopologyAwareHints.
+	topologyCache *topologycache.TopologyCache
+}
+
+// Run will not return until stopCh is closed.
+func (c *Controller) Run(ctx context.Context, workers int) {
+	defer utilruntime.HandleCrash()
+
+	// Start events processing pipeline.
+	c.eventBroadcaster.StartLogging(klog.Infof)
+	c.eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: c.client.CoreV1().Events("")})
+	defer c.eventBroadcaster.Shutdown()
+
+	defer c.queue.ShutDown()
+
+	logger := klog.FromContext(ctx)
```

Comment:
So this whole file is copy/pasted from https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpointslice/endpointslice_controller.go with very minor edits (https://github.com/cilium/cilium/pull/28440#issuecomment-1892021998) to remove import to `k8s.io/kubernetes/*`. Previous version of this PR just imported it through normal go dependency/vendor but it were adding ~160k LOC to the the vendor dir as a result (like it imported a lot of code from kube-controller-manager and kubelet packages, many things from otel...).

So technically I could change the logs here but it would mean other edits to this file and also klog is used throughout the rest of the reconciler code which I use as is without any edits through go dependency/vendor (this code https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/endpointslice/reconciler.go#L69) . So yeah not sure what we should do about the logs... FYI it does technically work and it logs stuff but in a different format as what Cilium uses unfortunately...

Reply:
> So this whole file is copy/pasted from

Oh I totally missed that. In that case it would make a lot of sense to fork endpointslice and have our patches on top of it.

I've created https://github.com/cilium/endpointslice-controller for this purpose, can we remove this duplicated code and make it available under https://github.com/cilium/endpointslice-controller? Then we can import https://github.com/cilium/endpointslice-controller into this repository.

Please ping me once you open a PR against https://github.com/cilium/endpointslice-controller

Thank you

Reply:
Thanks! Actually you already did the couple of modification that I needed to not get too much dependency so AFAIK we don't need anything else (at least for now?). But happy to followup and do some improvements in this area if needed ofc!